### PR TITLE
Add LLVM string type and global string variable definition function

### PR DESCRIPTION
**Pull request by rokopt**
This PR adds an LLVM string type which is implemented as an `i8*`, along with a function used to build literal strings by creating a global array of `i8` and then casting it to a pointer with `GEP`.

The code it's generating looks sane to me, but we can't return a `string` type from `main`, so to see it work we'll also need to add support for some imperative print routine such as `puts` or `printf` (or hand-edit the output program to add a call to one!).

### CODE REVIEW

**Comment by rokopt** [comment]
Sigh... the CI is failing with an indication that the affected LLVM golden test doesn't match the expected output, but there are no diffs between the expected and actual output (not even whitespace) that I (or `diff`) can find.

File: library/Backends/llvm/src/Juvix/Backends/LLVM/Compilation.hs [hunk 1]
```diff
@@ -195,6 +198,14 @@ compilePrimApp ty f xs
 -- | Write LLVM code for a primitive.
 -- TODO: implement other primitives.
 mkPrim ::
+  ( HasThrow "err" Types.Errors m,
+    HasState "blocks" (T Name.Name Types.BlockState) m,
```
Comment (Acentelles) [inline library/Backends/llvm/src/Juvix/Backends/LLVM/Compilation.hs:202]: Do we need all these constraints?

File: library/Backends/llvm/src/Juvix/Backends/LLVM/Codegen/Block.hs [hunk 2]
```diff
@@ -646,6 +647,34 @@ printf args = do
...
+        }
+  getElementPtr $
+    Types.Minimal
+      { Types.type' = Types.pointerOf Type.i8,
+        Types.address' =
+          ConstantOperand $
+            C.GlobalReference
+              (Types.pointerOf $ ArrayType (fromIntegral (length str + 1)) Type.i8)
+              name,
+        Types.indincies' = constant32List [0, 0]
```
Comment (Acentelles) [inline library/Backends/llvm/src/Juvix/Backends/LLVM/Codegen/Block.hs:675]:
```suggestion
        Types.indices' = constant32List [0, 0]
```

File: library/Backends/llvm/src/Juvix/Backends/LLVM/Parameterization.hs [hunk 3]
```diff
@@ -9,11 +9,24 @@ module Juvix.Backends.LLVM.Parameterization
   )
 where
 
+import qualified Juvix.Backends.LLVM.Codegen.Types as Types
 import Juvix.Backends.LLVM.Primitive
 import qualified Juvix.Core.Base.Types as Core
 import qualified Juvix.Core.IR.Evaluator as IR
 import qualified Juvix.Core.Parameterisation as Param
 import Juvix.Library
+  ( Applicative (pure),
```
Comment (Acentelles) [inline library/Backends/llvm/src/Juvix/Backends/LLVM/Parameterization.hs:18]: Is this intended?

File: library/Backends/llvm/src/Juvix/Backends/LLVM/Compilation.hs [hunk 4]
```diff
@@ -220,7 +235,7 @@ typeToLLVM ty@(ErasedAnn.Pi _usage _f _xs) =
   Types.pointerOf
     LLVM.FunctionType
       { LLVM.resultType = typeToLLVM resultType,
-        LLVM.argumentTypes = map typeToLLVM argumentTypes,
+        LLVM.argumentTypes = Juvix.Library.map typeToLLVM argumentTypes,
```
Comment (Acentelles) [inline library/Backends/llvm/src/Juvix/Backends/LLVM/Compilation.hs:238]: Can't we just use `map` or `<$>`

File: library/Backends/llvm/src/Juvix/Backends/LLVM/Codegen/Block.hs (hunk 2 above)
Comment (rokopt) [reply library/Backends/llvm/src/Juvix/Backends/LLVM/Codegen/Block.hs:675]: We could certainly make that change, but that spelling is in pre-existing code, so we can't make it without changing the other code which defines and uses it.  I'd suggest a separate PR for that, as it's a separate matter from supporting strings (a cosmetic one).

File: library/Backends/llvm/src/Juvix/Backends/LLVM/Parameterization.hs (hunk 3 above)
Comment (rokopt) [reply library/Backends/llvm/src/Juvix/Backends/LLVM/Parameterization.hs:18]: It was the pre-commit script that introduced it, so I think that if we didn't, then the script would just try to re-introduce into future PRs.  Maybe I'm misunderstanding something, though.

File: library/Backends/llvm/src/Juvix/Backends/LLVM/Compilation.hs (hunk 4 above)
Comment (rokopt) [reply library/Backends/llvm/src/Juvix/Backends/LLVM/Compilation.hs:238]: The new import of `HashMap` brought in another `map`, which is why I had to make the pre-existing one specify which `map` it was referring to.

File: library/Backends/llvm/src/Juvix/Backends/LLVM/Compilation.hs (hunk 1 above)
Comment (rokopt) [reply library/Backends/llvm/src/Juvix/Backends/LLVM/Compilation.hs:202]: Yes, the combination of the new `generateUniqueName` and the `moduleDefinitions` from `globalString` requires all of those.

File: library/Backends/llvm/src/Juvix/Backends/LLVM/Compilation.hs (hunk 4 above)
Comment (Acentelles) [reply library/Backends/llvm/src/Juvix/Backends/LLVM/Compilation.hs:238]: I think it's better to import HashMap qualified

**Review by Acentelles — APPROVED** [review]
Thanks for these changes 🎉

File: library/Backends/llvm/src/Juvix/Backends/LLVM/Compilation.hs (hunk 4 above)
Comment (mariari) [reply library/Backends/llvm/src/Juvix/Backends/LLVM/Compilation.hs:238]: We should import it qualified, we should also change the `map` call to `fmap`

File: library/Backends/llvm/src/Juvix/Backends/LLVM/Compilation.hs (hunk 1 above)
Comment (mariari) [reply library/Backends/llvm/src/Juvix/Backends/LLVM/Compilation.hs:202]: does the `Define` effect entail these, or should we make a more limited effect?

Comment (rokopt) [reply library/Backends/llvm/src/Juvix/Backends/LLVM/Compilation.hs:202]: Ah, good call -- `Define` does entail them all.  I'll substitute that.

File: library/Backends/llvm/src/Juvix/Backends/LLVM/Compilation.hs (hunk 4 above)
Comment (rokopt) [reply library/Backends/llvm/src/Juvix/Backends/LLVM/Compilation.hs:238]: I've now realized that the import was only required by an older version of the code, so I should simply remove it!

**Pull request merged**